Fix #3571: Introduce first version of Hints UI for the Exploration player

kevintab95 opened the pull request:
This PR introduces the new Hints UI for the Exploration player as a part of the Fallbacks to Hints project. Couple of things to note:
- Hints or Fallbacks need to be specified for card(s) to be able to use the UI in the player.
- Hint editor can be enabled by changing `oppia.constant('ENABLE_HINT_EDITOR', false);` to `oppia.constant('ENABLE_HINT_EDITOR', true);` in app.js to be able to specify Hints.
- The commits from migration PR #3555 has been merged with this which means Fallbacks are auto-migrated as Hints.

Known Issue:
If Fallbacks are specified instead of directly specifying Hints, they will get triggered if the answer is wrong 'x' times even if the Hint button wasn't clicked. This can either be corrected by removing the fallbacks UI completely or by removing the part where the trigger checks occur for now.

Linking to Issue  #3571.

Commit:
inital commit for hints and solution backend

Commit:
changed id to interaction_id

Commit:
fix linting error

Commit:
scrubbing html and minor changes

Commit:
separating views and failing tests

Commit:
fix comments and remove unnecessary html_cleaners

Commit:
remove 'hint view' related code

Commit:
fix linting error

Commit:
fix failing tests

Commit:
fix backend test failure

Commit:
migration of fallbacks to hints

Commit:
fix lint error

Commit:
fix for failing backend test

Commit:
Add comment for feedback[0]

Commit:
Introduce Hint editor frontend

Commit:
fix linting errors

Commit:
fix frontend errors

Commit:
Merged backend changes made in hints-solution-backend

Commit:
Merge remote-tracking branch 'origin/fallbacks-to-hints-migration' into hint-editor

Commit:
fallbacks to hints latest schema migration

Commit:
Merge remote-tracking branch 'origin/hint-editor' into fallbacks-to-hints-migration

Commit:
Fix third milestone

Commit:
fix linting error

Commit:
convert hint_text to hintText in the frontend

Commit:
Make suggested changes

Commit:
fix e2e errors

Commit:
change tmpHint to string type

Commit:
Suggested changes

Commit:
Remove one-off job

Commit:
Correct indentation in html files and alphabetizing imports

Commit:
Merged latest hint editor UI commit

Commit:
partial working of hints in learner view

Commit:
Disable Hint Editor

Commit:
Merge branch 'origin/hint-editor' into fallbacks-to-hints-migration

Commit:
Hints UI for inline and non-inline interactions

Commit:
Merge branch 'fallbacks-to-hints-migration' into hints-learner-ui

tjiang11 commented on core/templates/dev/head/pages/exploration_player/SupplementalCardDirective.js:
```diff
@@ -46,6 +46,13 @@ oppia.directive('supplementalCard', [
+            $scope.isCurrentHintUsable = false;
+            $scope.promiseForHint = $timeout(function () {
+              $scope.isCurrentHintUsable = true;
+            }, 10000);
```
Can you make this a named constant? Also, according to the design doc, this should be 30 seconds? Did you just use 10 for testing? 

Also, maybe you can factor these 4 lines out into it's own named function like 'resetHintTimer' as it's duplicated in the code. Up to you though

kevintab95 replied:
Yep, for testing. I've changed it to 30secs now.

tjiang11 commented on core/templates/dev/head/pages/exploration_player/SupplementalCardDirective.js:
```diff
@@ -62,6 +69,42 @@ oppia.directive('supplementalCard', [
+              $scope.isCurrentHintUsable = false;
+              $scope.promiseForHint = $timeout(function() {
+                $scope.isCurrentHintUsable = true;
+              }, 10000);
```
ditto--along with all the other instances of 10000. I'm wondering if it can be refactored such that we're not always duplicating a bunch of code across tutor and supplemental card, but that can be done at a later time.

kevintab95 replied:
Done.

tjiang11 commented on core/templates/dev/head/pages/exploration_player/SupplementalCardDirective.js:
```diff
@@ -62,6 +69,42 @@ oppia.directive('supplementalCard', [
+            var hints = $scope.currentInteractionHints;
+            if ($scope.hintButtonClicks < hints.length) {
+              var currentHint = hints[$scope.hintButtonClicks].hintText;
+              playerTranscriptService.addNewAnswer('(Asked for a hint)');
```
The text here seems pretty unnatural and awkward. Can you change it to something like "I would like a hint." We can change the wording more later if necessary, but I think '(Asked for a hint)' is really unfitting. Random idea for later: Maybe have a couple different variants like "I would like a hint", "I'm a little stuck, any hints?", "I'm having some trouble.", etc. just so it's not super repetitive and unnatural.

kevintab95 replied:
I've changed it to 'I would like a hint' for now. The variants idea seems nice, let's do that later!

tjiang11 commented on core/templates/dev/head/pages/exploration_player/supplemental_card_directive.html:
```diff
@@ -21,6 +21,15 @@
        angular-html-bind="activeCard.interactionHtml">
   </div>
+    <div ng-if="currentInteractionHints.length" style="text-align: right; padding: 8px">
+      <span tooltip="<[!isHintAvailable() ? 'Please try answering the question first.' : 'Click here for a hint!']>">
```
I know we don't actually want them to just wait, but with the current tooltip text I do get a really weird impression when the hint button suddenly becomes enabled when I didn't answer the question--like I take a moment to pause and ask myself "That's weird, it became enabled and I didn't answer the question". Maybe you can change the text to "Try thinking a bit more before asking for another hint!". 

> Once the Hint button is enabled, a pop-up tooltip will appear above the Hint button with the text “Click here for hints” to prompt the learner to try a Hint since he/she may be stuck at this point.

Also, my impression from the above quote in the design doc was that a tooltip will appear over the hint button once the timer times out, regardless of hover. Was this your intention?

kevintab95 replied:
Added text "Try thinking a bit more before asking for another hint!" for now. I did try the automatic tooltip/popover thing but it isn't working as expected, I'm not able to enable/disable tooltips based on events other that mouseover/click/focus events. Using JQuery would be an easy way to do this, but I'm trying to look for a more "angular"-style way of doing it.

tjiang11 commented on core/templates/dev/head/pages/exploration_player/SupplementalCardDirective.js:
```diff
@@ -62,6 +69,42 @@ oppia.directive('supplementalCard', [
+
+          $scope.isHintAvailable = function() {
+            var hintIsAvailable = (
+              $scope.isCurrentHintUsable && !$scope.allHintsExhausted);
```
You need to have some different text for when all the hints are exhausted, as it doesn't make sense to tell them to answer the question first when there are no hints remaining.

Actually, this should've been addressed earlier, but I just thought of this: maybe instead you should just completely hide the hint button if you're going to be showing the enabled solution button after all hints are exhausted. Because the hint button and solution button are never enabled simultaneously, right? So I think it would make sense to just show whichever one is enabled. And then I think it would make sense to hide both after the solution is shown.

kevintab95 replied:
I've added a new tooltip text that will say "Sorry, I'm out of hints!". Hmm, the disabling can be done though. Let's do it once the solution feature is in.

tjiang11 commented on core/templates/dev/head/pages/exploration_player/supplemental_card_directive.html:
```diff
@@ -21,6 +21,15 @@
+      <span tooltip="<[!isHintAvailable() ? 'Please try answering the question first.' : 'Click here for a hint!']>">
+        <md-button class="hint-button"
+                   ng-click="showHint()"
+                   ng-disabled="!isHintAvailable()">
```
I noticed that at one point I, being clumsy, accidentally tried submitting an answer by clicking the hint button instead of the submit button--it's a very tempting button to press as it's very dark/bold compared to the white submit button, and it's in a pretty convenient position, and it's in all CAPS which makes it stand out more (going to ask Mark if he has suggestions regarding the style of the button). Most people probably aren't as clumsy as I am, but I feel there will be instances where people make the same mistake. Maybe if the user has text in the response field, disable the hint button and have the tooltip be something like "Try submitting what you've typed first"

tjiang11 reviewed (changes requested):
Sorry for being late on this, was planning to do it last night but stuff came up. Anyways, it's generally looking really good!

I think we did mention before that the design on the hint button should be looked over, so tagging @markhalpin to take a look.

Also, can you rebase onto develop so that we have a clean PR? Right now this one includes a lot of stuff that has already been merged.

tjiang11 commented:
For reference @markhalpin 
![image](https://user-images.githubusercontent.com/11807091/27517520-ef608eb6-599b-11e7-92a3-a9d06b143aeb.png)
![image](https://user-images.githubusercontent.com/11807091/27517513-ce61e606-599b-11e7-8936-eb9c7c3770dc.png)
![image](https://user-images.githubusercontent.com/11807091/27517518-df7e8f52-599b-11e7-877d-e7a61a0f6bd0.png)

kevintab95 commented:
Please note, I've submitted another PR #3577 for the same issue. It is rebased on the latest develop branch in-order to have a cleaner PR. I've addressed review comments over there, so I'm closing this PR.
